### PR TITLE
[SELS-TASK] User Profile

### PR DESCRIPTION
**Pull request by louisepujante**
### Asana
-[link](https://app.asana.com/0/1201426075515354/1201470666211828/f)

### Commands
-cd backend > composer install > php artisan migrate:fresh --seed
-cd frontend > npm install > npm start

### Pre-conditions
-copy backend\ .env.example and rename to backend.env then provide local db credentials

viewing specific user profile:
-must be logged in and provide valid user credentials

### Expected Output
viewing specific user profile
-able to view specific user details (name, email, followers, following) and if current user is following this user or not

viewing followers and following:
-able to view profile of specific user's follower or following
-show if current user is following this user or not

### Screenshots:
viewing specific user profile:
image
image

viewing followers of specific user:
image

viewing following of specific user:
image

### Notes:
branch is based on feature/FE-BE-follow-feature, will rebase once it's merged

### CODE REVIEW

File: frontend/src/user/profiles/UserProfile.tsx [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+import { setIsError, setIsLoading } from '@store/category';
+
+interface SpecificUser {
+  id?: number;
+  name?: string;
+  user_image?: string;
+  email?: string;
+  followers?: FollowerRecord[];
+  following?: FollowingRecord[];
+}
```
Comment (inahsanjuan) [inline frontend/src/user/profiles/UserProfile.tsx:26]: May I know why this should be different?
Why not have a default `User` we can refer?

File: frontend/src/user/profiles/UserProfile.tsx [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+export default function UserProfile() {
+  const { user_id } = useParams();
+  const [cookies, _] = useCookies();
+  const dispatch = useDispatch();
+  const state = useSelector((state: RootState) => state);
+  const [user, setUser] = useState<SpecificUser>();
+  const [isViewingFollowers, setIsViewingFollowers] = useState(true);
+  const isFollowing = JSON.parse(
+    localStorage.getItem('user_following') || '{}'
```
Comment (inahsanjuan) [inline frontend/src/user/profiles/UserProfile.tsx:36]: Confirmation:
* Why are we manipulating the localStorage?

File: frontend/src/user/profiles/UserProfile.tsx [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        setUser({ ...user, following: user_following });
+      })
+      .catch();
+  };
+
+  const handleFollowUser = (user_id: number) => {
+    dispatch(setIsLoading(true));
+    const request = isFollowing
+      ? unfollowUser(cookies.user['id'], user_id)
+      : followUser(cookies.user['id'], { following_id: user_id });
```
Comment (inahsanjuan) [inline frontend/src/user/profiles/UserProfile.tsx]: This kind of condition sometimes may not be secure.
Please create totally different handler for following and unfollowing, and call these function directly in the rendered jsx if needed

File: frontend/src/user/profiles/UserProfile.tsx [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        <>
+          {state.category.isLoading && !state.category.isLoading && <Loader />}
+          {state.category.isError && (
+            <Notification
+              isSuccess={false}
+              title='An error has occurred. Please try again later.'
+              errorAction='refresh'
+            />
+          )}
+          {!state.category.isLoading && !state.category.isLoading && (
```
Comment (inahsanjuan) [inline frontend/src/user/profiles/UserProfile.tsx]: Some conditions are hard to follow, can you put comments on explaining why this is being rendered like this?

File: frontend/src/user/profiles/UserProfileItem.tsx [hunk 5]
```diff
@@ -1,60 +1,72 @@
...
-    return following.user_id === props.user.id;
+  const user_following = JSON.parse(
+    localStorage.getItem('user_following') || '{}'
+  );
+  const isFollowing = user_following.find((following: FollowingRecord) => {
+    return following.following[0].id === props.user.id;
   });
 
-  const handleFollowUser = () => {
+  const handleFollowUser = () => {
```
Comment (inahsanjuan) [inline frontend/src/user/profiles/UserProfileItem.tsx]:
```suggestion
  const handleFollowUser = () => {
```

**Review by inahsanjuan — CHANGES_REQUESTED** [review]

File: frontend/src/user/profiles/UserProfile.tsx (hunk 1 above)
Comment (louisepujante) [reply frontend/src/user/profiles/UserProfile.tsx:26]: this is because the response for retrieving the followers/following is different miss.

User response looks like:
```
 {
  id: 2;
  name: 'Name here';
  email: 'email@email.com';
  user_image: 'null';
}
```


but the followers/following response looks like:
```
{
  "id": 16,
  "user_id": 2,
  "following": [
      {
          "id": 5,
          "name": "Name of user that user2 is following",
          "email": "email@email.com",
          "user_image": null
      }
   ]
}
```

File: frontend/src/user/profiles/UserProfile.tsx (hunk 3 above)
Comment (louisepujante) [reply frontend/src/user/profiles/UserProfile.tsx]: noted miss, will change

File: frontend/src/user/profiles/UserProfile.tsx (hunk 2 above)
Comment (louisepujante) [reply frontend/src/user/profiles/UserProfile.tsx:36]: i'm storing the retrieved user_following (users that currently logged in user is following) because when I tried to store it in the state, the values are gone when I refresh the page. the user_following value is used to indicate if the followers/following of another user is also being followed by the current user
image

value is also used here, when current user is viewing another user's profile
image

File: frontend/src/user/profiles/UserProfile.tsx (hunk 4 above)
Comment (louisepujante) [reply frontend/src/user/profiles/UserProfile.tsx]: okay miss, will add comments

File: frontend/src/user/profiles/UserProfileItem.tsx (hunk 5 above)
Comment (louisepujante) [reply frontend/src/user/profiles/UserProfileItem.tsx]: noted miss, will remove
